src/library/browse/browsefeature.cpp

```diff
@@ -400,28 +401,47 @@ void BrowseFeature::onLazyChildExpandation(const QModelIndex& index) {
     if (path == DEVICE_NODE) {
         folders += getRemovableDevices();
     } else {
-        // we assume that the path refers to a folder in the file system
-        // populate childs
-        MDir dir(path);
-
-        QFileInfoList all = dir.dir().entryInfoList(
-            QDir::Dirs | QDir::NoDotAndDotDot);
-
-        // loop through all the item and construct the childs
-        foreach (QFileInfo one, all) {
-            // Skip folders that end with .app on OS X
+        auto listWorker = [=]() mutable -> void {
+            // we assume that the path refers to a folder in the file system
+            // populate childs
+            MDir dir(path);
+            QList<TreeItem*> children;
+            QFileInfoList all = dir.dir().entryInfoList(
+                    QDir::Dirs | QDir::NoDotAndDotDot);
+
+            // loop through all the item and construct the childs
+            foreach (QFileInfo one, all) {
 #if defined(__APPLE__)
-            if (one.isDir() && one.fileName().endsWith(".app"))
-                continue;
+                if (one.isDir() && one.fileName().endsWith(".app"))
+                    continue;
 #endif
-            // We here create new items for the sidebar models
-            // Once the items are added to the TreeItemModel,
-            // the models takes ownership of them and ensures their deletion
-            TreeItem* folder = new TreeItem(
-                one.fileName(),
-                QVariant(one.absoluteFilePath() + QStringLiteral("/")));
-            folders << folder;
-        }
+                // We here create new items for the sidebar models
+                // Once the items are added to the TreeItemModel,
+                // the models takes ownership of them and ensures their deletion
+                auto* folder = new TreeItem(
+                        one.fileName(),
+                        QVariant(one.absoluteFilePath() + QStringLiteral("/")));
+                children << folder;
+            }
+            if (!children.isEmpty()) {
+                m_childModel.insertTreeItemRows(children, 0, index);
```

Comment:
Accessing members from another thread is unsafe and not allowed.

https://doc.qt.io/qt-5/threads-qobject.html#accessing-qobject-subclasses-from-other-threads

> QObject and all of its subclasses are not thread-safe.

Reply:
hence the comment in the PR description to protect the data/hashmap of `m_childModel` with mutex.

Reply:
A mutex is not the right tool here. Use signals/slots for communication between threads. Otherwise all access to m_childModel in the UI thread would need to be protected. This would spread like cancer.

Reply:
sure; but note that currently signal/slot in `main` is totally overwhelmed for some reason (spinny, vumeter and sidebar) it's just taking more than the cpu can handle... but i guess this is a regression that'll be fixed in upstream

Reply:
If you want to talk to a QObject that is driven by an event loop *emit* signals (that are connected to its slots) or *invoke* slots directly.

Reply:
yeah sure i'm not questioning that; just saying that currently the app is in an unusable state in `main` and that is due to the overhead created with signals/slots

Reply:
unusable state: UI is blocking and super laggy when playing 2 decks... need to run it in `--safeMode` etc...

Reply:
I don't deny that the current state of the browse view is far from optimal and needs to be improved. Nevertheless, introducing code that is incorrect, will cause data inconsistencies and probably crashes is not an option.

Reply:
Does this actually work? I am asking because everything is captures by value [=] this means If I am right, you user here a copy of m_childModel.

Reply:
Ah no, you are using a copy of this and it is actually this->m_childModel, right?

Reply:
> I don't deny that the current state of the browse view is far from optimal and needs to be improved. Nevertheless, introducing code that is incorrect, will cause data inconsistencies and probably crashes is not an option.

:) maybe we misunderstood each other somewhere. i never intended not to not (triple negative?) include your suggestion to fix the patch 🙃 and thnx for catching that! 👍 just mentioned the reason why i stayed away from signaling/slots... doesn't mean that it was the correct way, but i just wanted to get rid of the jitter of UI and minimize the call overhead. note unfortunately this was not actually solving all the jitter - and the reason for that is totally unclear for me :(